README.md
## Overview ##

Lua for Windows is a 'batteries included environment' for the Lua scripting language on Windows.

Lua for Windows (LfW) combines Lua binaries, Lua libraries with a Lua-capable editor in a single install package for the Microsoft Windows operating system. LfW contains everything you need to write, run and debug Lua scripts on Windows. A wide variety of libraries and examples are included that are ready to use with Microsoft Windows. LfW runs on Windows 2000 and newer versions of Windows. Lua and its associated libraries are also available for other operating systems, so most scripts will be automatically cross-platform.


## Download ##

Download Lua For Windows installer from [GitHub](http://github.com/rjpcomputing/luaforwindows/releases)

## Install Notes ##

Lua for Windows installs Lua language, SciTE based Lua IDE and Lua modules to the directory of you choice at install time. Lua for Windows and it's modules all depend on the MSVC++ 2005 runtime library. Lua for Windows install will automatically download this runtime and install it for you if you don't have runtime installed on your computer. The runtime is contained in the file [vcredist\_x86.exe](http://github.com/rjpcomputing/luaforwindows/releases/download/vc-runtime/vcredist_x86.4053.exe), if you don't want this download to occur at installation then place the [vcredist\_86.exe](http://github.com/rjpcomputing/luaforwindows/releases/download/vc-runtime/vcredist_x86.4053.exe) in same directory as Lua for Windows install exe.


## LfW Components ##

Installer
Lua Interpreter,
Lua Reference Manual,
Quick Lua Tour Sample,
Examples directory,
Libraries with documentation.
Lua Compiler (luac)
Text Editors (SciTE and I think maybe wxLua editors too)
C header files/libraries/etc. for building C module


## Libraries included ##
|**Library**|**Version**|**Description**|
|:----------|:----------|:--------------|
|[Alien](http://alien.luaforge.net/)|0.5.0|Provides access to functions in an unknown or new .dll.|
|[IUP](http://www.tecgraf.puc-rio.br/iup/)|3.5.0|Light Portable Graphical User Interface library.|
|[CD](http://www.tecgraf.puc-rio.br/cd/)|5.4.1|Canvas Draw: A platform-independent graphic library.|
|[IM](http://www.tecgraf.puc-rio.br/im/)|3.6.3|A toolkit for Digital Imaging.|
|Ex|Jan 07|Adds environment, file system, I/O (Locking and pipes), and process control.|
|[LPeg](http://www.inf.puc-rio.br/~roberto/lpeg/lpeg.html)|0.9|Pattern-matching library based on Parsing Expression Grammars (PEGs).|
|[Lua-GD](http://lua-gd.luaforge.net)|2.9.33r2|Image manipulation library based on Thomas Boutell’s GD library.|
|[LuaCOM](http://www.tecgraf.puc-rio.br/~rcerq/luacom/)|1.4|Enable use & implementation of Microsoft’s Component Object Model.|
|[LuaCURL](http://luacurl.luaforge.net/)|1.0|Interface to Internet browsing capabilities based on the cURL library.|
|[Date](http://luaforge.net/projects/date/)|2 |Date and Time library for Lua.|
|[LuaDoc](http://luadoc.luaforge.net/)|3.01|Documentation tool for Lua source code.|
|[LuaExpat](http://www.keplerproject.org/luaexpat/)|1.1.0|Lua interface to XML Expat parsing library.|
|[LuaFileSystem](http://keplerproject.github.com/luafilesystem/)|1.4.2|Access the directory structure and file attributes.|
|[LuaLogging](http://www.keplerproject.org/lualogging/)|1.2.0|Logging features in Lua, based on log4j.|
|[LuaProfiler](http://luaprofiler.luaforge.net/manual.html)|2.0.1|Time profiler designed to find bottlenecks in Lua programs.|
|[LuaSocket](http://www.tecgraf.puc-rio.br/~diego/professional/luasocket/)|2.0.2|Lua interface to support HTTP,FTP,SMTP, MIME, URL & LTN12.|
|[LuaSQL](http://www.keplerproject.org/luasql/)|2.1.1|Lua interface for PostgreSQL, ODBC, MySQL, SQLite, Oracle, and ADO dbms.|
|[LuaUnit](https://github.com/luaforge/luaunit)|2.0|Testing framework for Lua.|
|[LuaZip](http://www.keplerproject.org/luazip/)|1.2.3|Read files from zip files.|
|[stdlib](http://luaforge.net/projects/stdlib/)|25|Library of modules for common programming tasks, including list, table and functional operations, regexps, objects, pretty-printing and getopt.|
|[lrexlib](http://math2.org/luasearch/rex.html#contact)|2.2|Regular expression library for Lua.|
|[MD5](http://www.keplerproject.org/md5/)|1.1.2|Basic cryptographic facilities for Lua.|
|[Copas](http://www.keplerproject.org/copas/)|1.1.5|Dispatcher based on coroutines that can be used by TCP/IP servers.|
|[Coxpcall](http://coxpcall.luaforge.net/)|1.13|Encapsulates the protected calls with a coroutine based loop, so errors can be dealed without the usual pcall/xpcall issues with coroutines.|
|[Rings](http://www.keplerproject.org/rings/)|1.2.2|Provides a way to create new Lua states from within Lua. It also offers a simple way to communicate between the creator (http://loop.luaforge.net/) and the created (slave) states.|
|[LOOP](http://loop.luaforge.net/)|2.3 Beta|LOOP stands for Lua Object-Oriented Programming and is a set of packages for supporting different models of object-oriented programming in the Lua language.|
|[LuaTask](http://luatask.luaforge.net/)|1.6.4|Implements a concurrent and independent Lua execution environment model.|
|[LuaInterface](http://code.google.com/p/luainterface/)|1.5.3|	Integration between the Lua language and Microsoft .NET platform's Common Language Runtime (CLR).|
|[wxLua](http://wxlua.sourceforge.net/)|2.8.10| Lua binding to the wxWidgets library.|
|[lpack](http://www.tecgraf.puc-rio.br/~lhf/ftp/lua/#lpack)|29 Jun 2007 19:27:20|Lua library for packing and unpacking binary data.|
|[VStruct](http://www.funkyhorror.net/toxicfrog/projects/vstruct/)|1.0.2|Provides functions for manipulating binary data.|
|[LuaBitOps](http://bitop.luajit.org/)|1.0.1| Lua BitOp is a C extension module for Lua 5.1 which adds bitwise operations on numbers.|
|[LuaXML](http://viremo.eludi.net/LuaXML/)|1.3|A module that maps between Lua and XML without much ado.|
|[Lanes](http://luaforge.net/projects/lanes/)|2.0.4|Provides the possibility to run multiple Lua states in parallel.|
|[MetaLua](http://metalua.luaforge.net/)|0.5-rc1|A complete macro system with full compatibility with Lua 5.1 sources and bytecode: clean, elegant semantics and syntax, amazing expressive power, good performances, near-universal portability.|
|[LuaGL](http://luagl.wikidot.com/)|1.3|Provides access to all of the OpenGL functionality from Lua 5.x .|
|[Penlight](http://stevedonovan.github.com/Penlight/)|1.3.2|Common lua code patterns for tables, arrays, strings, paths and directories, data, and functional programming.|
|[lbase64](http://www.tecgraf.puc-rio.br/~lhf/ftp/lua/)|for Lua 5.1|A base64 library for Lua.|
|[gzio](http://github.com/jmaygarden/gzio)|0.9.0|The Lua gzip file I/O module emulates the standard I/O module, but operates on compressed gzip format files.|
|[LuaRS232](http://github.com/ynezz/librs232)|1.0.0| RS232 serial communications library|
|[LeMock](http://lemock.luaforge.net/README.html)|0.6|  Mock creation module intended for use together with a unit test framework such as lunit or lunity.|
|[LuaRocks](http://luarocks.org)|2.0.2|  A deployment and management system for Lua modules.|
|[Oil](http://oil.luaforge.net)|0.4-beta|  It is a simple, efficient and flexible object request broker written in the Lua language.|
|[LuaJSON](http://github.com/harningt/luajson)|1.2.2| JSON parser/encoder for Lua Parses JSON using LPEG for speed and flexibility. Depending on parser/encoder options, various values are preserved as best as possible.|
|[SubLua](https://sourceforge.net/projects/subcpp/)|1.8.10-35352|  Subversion binding.|
|[30Log](https://github.com/Yonaba/30log)|1.0.0|  30 lines library for object orientation in Lua.|
|[Moses](http://yonaba.github.io/Moses)|1.4.0|  Utility library for functional programming in Lua.|
|[Binary-Heaps](http://yonaba.github.com/Binary-Heaps)|1.5.1|  Implementation of binary heaps data structure in pure Lua.|
|[strictness](http://yonaba.github.io/strictness)|0.2.0|  Tracks accesses and assignments of undefined variables in Lua.|
|[delaunay](http://yonaba.github.io/delaunay/)|0.1.0|  Delaunay triangulation for convex polygon.|

## History ##

**5.1.3.12** 9/July/2008 - First release

**5.1.3.13** 16/July/2008 - Second release
  * New libs LOOPS, LuaTask, LuaInterface.
  * Add Lua headers and Lua lib needed for embedding Lua.

**5.1.3.15** 7/August/2008 - Third release
  * Added an environment variable called "LUA\_DEV" to help C module developers configure their build files. This allows for the paths to be common and easily added to the project settings. (e.g. $(LUA\_DEV)/include; $(LUA\_DEV)/lib)
  * Added the ability to download the 'vcredist\_x86.exe'.
  * Fixed Visual C++ 8.0 run-time problems. There now has to be a run-time for every dll loaded.
  * Removed 'msvcr80.dll' from the root application directory.
  * Updated the Visual C++ 8.0 run-time to SP1.
  * Updated StdLib to version 6.
  * Fixed missing Scite install issue with 5.1.3.14 release.

**5.1.4.17** 9/September/2008 - Fourth release.
  * Fixed the readyiss.bat file to work with the new 'lua' directory.
  * Added wxLua support.
  * Updated to Lua v5.1.4.
  * Changed the 'lualibs' to 'lua' to support the default locations.
  * Updated SciTE property file to make the help show up correctly.
  * Removed setting the LUA\_PATH and LUA\_CPATH environment variables.

**5.1.4.18** 11/September/2008 - Fifth release.
  * Updated Stdlib to version 8.
  * Updated the quickluatour.lua.
  * Updated the luaforwindows document.
  * Fixed LuaInterface module.

**5.1.4.19** 08/October/2008 - 6th release.
  * Release on  http://luaforwindows.googlecode.com/ to cover when LuaForge site was unavailable.

**5.1.4.20**  26/November/2008 - 7th release.
  * Added lpack (29 Jun 2007 19:27:20) module.
  * Added VStruct v 1.0 beta2 Lua module.
  * Added BitLib release 25 module.
  * Added LuaXML v1.3 module.
  * Updated IUP to version 2.7.
  * Updated CD to version 5.1.
  * Updated IM to version 3.4.

**5.1.4.21** 5/December/2008 - 8th release.
  * Fixed a SciTE bug when opening the Lua manual.
  * Added lExecutor. This is a wxLua application that allows users to run console Lua scripts in a simple GUI. Look at the documentation (<lFW root>\docs\lexecutor) for more information.
  * Added 'wildcard'.lexe files to run using lExecutor.
  * Associated 'wildcard'.lexe files to a custom icon.

**5.1.4.22** 12/January/2009- 9th release.
  * Added LuaBitOp v1.0.1 library.
  * Fixed a bug in lExecutor to properly pass file loading error to the user.
  * Fixed LuaDoc templates a bit for 'tables' and 'functions'.
  * Fixed !LuaXML so it actually works. Opps.
  * Updated Copas to version 1.1.4.
  * Updated IUP to version 2.7.1.
  * Updated CD to version 5.1.1.
  * Updated IM to version 3.4.1.
  * Removed BitLib. Use LuaBitOp instead.

**5.1.4.23** 06/Febuary/2009 - 10th release.
  * Added MetaLua v0.5-rc1.
  * Added the 'strict.lua' file shipped in the Lua in the 'etc' directory.
  * Added Lanes v2.0.3 library.
  * Updated LuaFileSystem to version 1.4.2.

**5.1.4.24** **rc1** 14/April/2009 - 11th release
  * Updated LPeg to version 0.9.
  * Updated Copas to version 1.1.5.
  * Updated the Lua executables so they run on Windows 64

**5.1.4.25** **rc2** 15/April/2009 - 12th release
  * Updated lua51.dll manifest to allow running on Windows 64.

**5.1.4.26** **rc3** 23/April/2009 - 13th release
  * Added an option to start the quickluatour.lua after installation completes.
  * Added a link to the lExecutor documentation in the start menu.
  * Now you can launch quickluatour.lua at the end of the installer
  * Updated luaforwindows.html to be the right version in the example.
  * Updated the LuaQuickTour thanks to Dirk Feytons.
  * Fixed SciTE saving a debugger file to the current running scripts directory. Now it saves it to your local temp directory.
  * Fixed SciTE toolbar icons being incorrect.

**5.1.4-27** **rc4** 27/April/2009 -14th release
  * Fixed SciTE debugger looking for the temp file in the %TMP% location.
  * Fixed debugging files with spaces in the filename.
  * Updated the IM, CD, and IUP examples.
  * Made SciTE use the lua5.1.dll.
  * Changed the version number to use a '-' as a build number of the LfW package.
  * Changed the installer display name to 'Lua For Windows'.

**5.1.4-28** 29/June/2009 - 15th release
  * Fixed SciTE debugger to allow it to debug modules in the same directory  as the file using the module.
  * Updated Stdlib to version 11.

**5.1.4-29** 03/Sept/2009 - 16th release
  * Added Penlight v0.7.2 Lua module.
  * Updated LuaCOM to version 1.4.
  * Updated LuaGL to version 1.3.
  * Updated VStruct to version 1.0.2.

**5.1.4-30** 10/Sept/2009 - 17th release
  * Added lbase64 module.
  * Added gzio module
  * Updated stdlib to Release 12.

**5.1.4-31** 07/Dec/2009 -  Internal only, not released
  * Added markdown v0.32 Lua module.
  * Fixed C Runtime Error R 6034. Problem 3.
  * Fixed LuaInterface manifest loading issue.
  * Updated luanet.dll to CoInitialize and search in the LUA\_CPATH.

**5.1.4-32** 12/March/2010 - 18th release
  * Added LuaRS232 v1.0.0
  * Added iLua.cmd to easily launch iLua. Also added it to the start menu.
  * Fixed SciTE to properly run interactive Lua session.
  * Updated the clidebugger to take the same paramaters to allow for debug level.
  * Updated Penlight to version 0.8.0.
  * Updated IUP to version 3.0.0.
  * Updated CD to version 5.3.0.
  * Updated IM to version 3.6.0.

**5.1.4-33** 12/March/2010 - 19th release
  * Added LeMock v0.6

**5.1.4-34** 18/March/2010 - 20th release
  * Updated Penlight to version 0.8.1

**5.1.4-35** 07/April/2010 - 21th release
  * Updated lExecutor to version 1.01
  * Updated Penlight to version 0.8.2

**5.1.4-36** 09/April/2010 - 22th release
  * Updated Penlight to version 0.8.3

**5.1.4-37** 04/May/2010 - 23th release
  * Updated Penlight to version 0.8.4

**5.1.4-38** 18/May/2010 - 24th release
  * Fixed #10 - LuaDoc incomplete. Added luadoc\_start.bat.
  * Fixed #09  - luanet kill lua.exe
  * Updated the SciTE .api file. Now includes lfs, bit and pack.

**5.1.4-39** 17/June/2010 - 25th release
  * Added LuaRocks version 2.0.2.
  * Updated wxLua to version 2.8.10.
  * Updated stdlib to release 15.
  * Updated stdlib to release 14.
  * Updated stdlib to release 13.

**5.1.4-40** 26/July/2010 - 26th release
  * Fixes #14 - Just a simple mistake in the link to LuaEx.
  * Fixes #18 - Lua tutorial typo.
  * Reverted to wxLua version 2.8.7 because there are issues with the new versions stability. It will get updated later.
  * Fixed pplot examples that ship with IUP.
  * Closed #20 - Updated IUP to version 3.2.0.
  * Updated Lanes to version 2.0.4.
  * Added Oil version 0.4-beta.

**5.1.4-41** 11/Jan/2011  - 27th release
  * Fixes #34 - LuaXML clib delivered with LuaForWindows causes memory leaks.

**5.1.4-42** 01/June/2011 - 28th release
  * Updated LuaLogging to version 1.2.0.
  * Updated stdlib to release 20.
  * Updated Penlight to version 0.9.4.

**5.1.4-43** 06/June/2011 - 29th release
  * Fixes #1 - LuaSocket example not upto date - cddb.lua.
  * Fixes #24 - Commented-out visible whitespace acquires coment colour.
  * Closed #19 - Added LuaJSON v1.2.2 library.
  * Closed #41 - Updated Alien to version 0.5.0.
  * Updated LPeg to version 0.10-1.
  * Updated stdlib to release 21.

**5.1.4-44** 08/June/2011 - 30th release
  * Fixed pl.dir so you can use strict module

**5.1.4-45** 14/June/2011 - 31st release
  * Fixes #42 - Please update Tecgraf libs (CD, IM, IUP)
  * Updated IUP to version 3.5.0.
  * Updated CD to version 5.4.1.
  * Updated IM to version 3.6.3.

**5.1.4-46** 07/August/2012 - 32nd release
  * Fixes #43 - require('lpeg') -- system error 14001
  * Updated stdlib to release 25.
  * Updated the Visual C++ 8.0 run-time to SP1-4053.
  * Updated lpeg.dll manifest to allow running on older vcredist\_x86.
  * Updated alien/core.dll manifest to allow running on older vcredist\_x86.
  * Updated alien/struct.dll manifest to allow running on older vcredist\_x86.
  * Updated examples/alien/tests/alientest.dll manifest to allow running on older vcredist\_x86.
  * Updated LuaXML\_lib.dll manifest to allow running on older vcredist\_x86.
  * Updated w32.dll manifest to allow running on older vcredist\_x86.
  * Updated Penlight to version 1.0.3.
  * Updated LuaUnit to version 2.0.
  * Added SubLua version 1.7.4 - Subversion binding

**5.1.4-47** 18/March/2015 - 33rd release
  * Updated stdlib to release 28.
  * Updated Penlight to 1.3.2.
  * Updated SubLua to 1.8.10.
  * **WARNING**: Moved all downloads and code hosting to GitHub. Older releases will not function when Google Code is shut down. Make sure to upgrade.

**5.1.4-48** 16/September/2015 - 34rd release
  * Added 30Log v1.0.0 - 30 lines library for object orientation in Lua
  * Updated SubLua to 1.8.11.

**5.1.4-49** 23/September/2015 - 35th release
  * Added delaunay Lua module
  * Added Binary Heaps Lua module
  * Added Moses Lua module

**5.1.4-50** 22/October/2015 - 36th release
  * Fixes #98 - Lua For Windows Installer - 'h' is an invalid URL.
  * Changed app URL to point to GitHub.
  * Changed minimum version to reflect the GitHub version.

**5.1.5-51** 29/March/2016 - 37th release
  * Updated to Lua 5.1.5.
  * Updated SubLua to 1.8.10-35352


changes.txt
=================== Lua For Windows ===================
08/15/2017
* Disabling vcredist_x86 download to prevent installation issues.

03/29/2016  Version 5.1.5-51
^ Updated Lua to 5.1.5.
^ Updated SubLua to 1.8.10-35352.

10/22/2015  Version 5.1.4-50
* Fixes #98 - Lua For Windows Installer - 'h' is an invalid URL.
* Changed app URL to point to GitHub.
* Changed minimum version to reflect the GitHub version.

09/23/2015  Version 5.1.4-49
+ Added delaunay Lua module
+ Added Binary Heaps Lua module
+ Added Moses Lua module

09/16/2015  Version 5.1.4-48
+ Added 30Log v1.0.0 - 30 lines library for object orientation in Lua
^ Updated SubLua to 1.8.10.

03/18/2015  Version 5.1.4-47
^ Updated stdlib to release 28.
^ Updated Penlight to 1.3.2.
^ Updated SubLua to 1.8.10.
* Moved all downloads and code hosting to GitHub. Older
  releases will not function when Google Code is shut down.
  Make sure to upgrade.

08/07/2012  Version 5.1.4-46
* Fixes #43 - require('lpeg') -- system error 14001
^ Updated stdlib to release 25.
^ Updated the Visual C++ 8.0 run-time to SP1-4053.
^ Updated lpeg.dll manifest to allow running on older vcredist_x86.
^ Updated alien/core.dll manifest to allow running on older vcredist_x86.
^ Updated alien/struct.dll manifest to allow running on older vcredist_x86.
^ Updated examples/alien/tests/alientest.dll manifest to allow running on older vcredist_x86.
^ Updated LuaXML_lib.dll manifest to allow running on older vcredist_x86.
^ Updated w32.dll manifest to allow running on older vcredist_x86.
^ Updated Penlight to version 1.0.3.
^ Updated LuaUnit to version 2.0.
+ Added SubLua version 1.7.4 - Subversion binding

06/14/2011  Version 5.1.4-45
* Fixes #42 - Please update Tecgraf libs (CD, IM, IUP)
^ Updated IUP to version 3.5.0.
^ Updated CD to version 5.4.1.
^ Updated IM to version 3.6.3.

06/08/2011  Version 5.1.4-44
* Fixed pl.dir so you can use strict module

06/06/2011  Version 5.1.4-43
* Fixes #1 - LuaSocket example not upto date - cddb.lua.
* Fixes #24 - Commented-out visible whitespace acquires coment colour.
+ Closed #19 - Added LuaJSON v1.2.2 library.
^ Closed #41 - Updated Alien to version 0.5.0.
^ Updated LPeg to version 0.10-1.
^ Updated stdlib to release 21.

06/01/2011  Version 5.1.4-42
^ Updated LuaLogging to version 1.2.0.
^ Updated stdlib to release 20.
^ Updated Penlight to version 0.9.4.

01/11/2011  Version 5.1.4-41
* Fixes #34 - LuaXML clib delivered with LuaForWindows causes memory
  leaks.

07/26/2010  Version 5.1.4-40
* Fixes #14 - Just a simple mistake in the link to LuaEx.
* Fixes #18 - Lua tutorial typo.
* Reverted to wxLua version 2.8.7 because there are issues with the new
  versions stability. It will get updated later.
* Fixed pplot examples that ship with IUP.
^ Closed #20 - Updated IUP to version 3.2.0.
^ Updated Lanes to version 2.0.4.
+ Added Oil version 0.4-beta.

06/17/2010  Version 5.1.4-39
+ Added LuaRocks version 2.0.2.
* Fixes #13 - Updated wxLua to version 2.8.10.
^ Updated stdlib to release 15.
^ Updated stdlib to release 14.
^ Updated stdlib to release 13.

05/18/2010  Version 5.1.4-38
* Fixed #10 - LuaDoc incomplete. Added luadoc_start.bat.
* Fixed #09 - luanet kill lua.exe
^ Updated the SciTE .api file. Now includes lfs, bit and pack.

05/04/2010  Version 5.1.4-37
^ Updated Penlight to version 0.8.4

04/09/2010  Version 5.1.4-36
^ Updated Penlight to version 0.8.3

04/07/2010  Version 5.1.4-35
^ Updated lExecutor to version 1.01
^ Updated Penlight to version 0.8.2

03/18/2010  Version 5.1.4-34
^ Updated Penlight to version 0.8.1

03/15/2010  Version 5.1.4-33
+ Added LeMock v0.6

03/12/2010  Version 5.1.4-32
+ Added LuaRS232 v1.0.0
+ Added iLua.cmd to easily launch iLua. Also added it to the start menu.
* Fixed SciTE to properly run interactive Lua session.
^ Updated the clidebugger to take the same paramaters to allow for debug level.
^ Updated Penlight to version 0.8.0.
^ Updated IUP to version 3.0.0.
^ Updated CD to version 5.3.0.
^ Updated IM to version 3.6.0.

12/07/2009  Version 5.1.4-31
+ Added markdown v0.32 Lua module.
* Fixed Issue 4: C Run-Time Error R6034.
* Fixed LuaInterface manifest loading issue.
^ Updated luanet.dll to CoInitialize and search in the LUA_CPATH.

09/15/2009  Version 5.1.4-30
+ Added lbase64 version for 5.1 module.
+ Added gzio version 0.9.0 module.
^ Updated Stdlib to version 12.
^ Updated the LfW documentation to describe how to add your own modules
  to the system.

09/02/2009  Version 5.1.4-29
+ Added Penlight v0.7.2 Lua module.
^ Updated LuaCOM to version 1.4.
^ Updated LuaGL to version 1.3.
^ Updated VStruct to version 1.0.2.

06/29/2009  Version 5.1.4-28
* Fixed SciTE debugger to allow it to debug modules in the same directory
  as the file using the module.
^ Updated Stdlib to version 11.

04/27/2009  Version 5.1.4-27 rc4
* Fixed SciTE debugger looking for the temp file in the %TMP% location.
* Fixed debugging files with spaces in the filename.
^ Updated the IM, CD, and IUP examples.
^ Made SciTE use the lua5.1.dll.
^ Changed the version number to use a '-' as a build number of the LfW
  package.
^ Changed the installer display name to 'Lua For Windows'.

04/23/2009  Version 5.1.4.26 rc3
+ Added an option to start the quickluatour.lua after installation
  completes.
+ Added a link to the lExecutor documentation in the start menu.
+ Now you can launch quickluatour.lua at the end of the installer
^ Updated luaforwindows.html to be the right version in the example.
^ Updated the LuaQuickTour thanks to Dirk Feytons.
* Fixed SciTE saving a debugger file to the current running scripts
  directory. Now it saves it to your local temp directory.
* Fixed SciTE toolbar icons being incorrect.

04/15/2009  Version 5.1.4.25 rc2
^ Updated lua51.dll manifest to allow running on Windows 64.

04/14/2009  Version 5.1.4.24 rc1
^ Updated LPeg to version 0.9.
^ Updated Copas to version 1.1.5.
^ Updated the Lua executables so they run on Windows 64.

02/06/2009  Version 5.1.4.23
+ Added MetaLua v0.5-rc1.

files/lua/re.lua
-- $Id: re.lua,v 1.38 2010/11/03 17:21:07 roberto Exp $

-- imported functions and modules
local tonumber, type, print, error = tonumber, type, print, error
local setmetatable = setmetatable
local m = require"lpeg"

-- 'm' will be used to parse expressions, and 'mm' will be used to
-- create expressions; that is, 're' runs on 'm', creating patterns
-- on 'mm'
local mm = m

-- pattern's metatable 
local mt = getmetatable(mm.P(0))



-- No more global accesses after this point
local version = _VERSION
if version == "Lua 5.2" then _ENV = nil end


local any = m.P(1)


-- Pre-defined names
local Predef = { nl = m.P"\n" }


local mem
local fmem
local gmem


local function updatelocale ()
  mm.locale(Predef)
  Predef.a = Predef.alpha
  Predef.c = Predef.cntrl
  Predef.d = Predef.digit
  Predef.g = Predef.graph
  Predef.l = Predef.lower
  Predef.p = Predef.punct
  Predef.s = Predef.space
  Predef.u = Predef.upper
  Predef.w = Predef.alnum
  Predef.x = Predef.xdigit
  Predef.A = any - Predef.a
  Predef.C = any - Predef.c
  Predef.D = any - Predef.d
  Predef.G = any - Predef.g
  Predef.L = any - Predef.l
  Predef.P = any - Predef.p
  Predef.S = any - Predef.s
  Predef.U = any - Predef.u
  Predef.W = any - Predef.w
  Predef.X = any - Predef.x
  mem = {}    -- restart memoization
  fmem = {}
  gmem = {}
  local mt = {__mode = "v"}
  setmetatable(mem, mt)
  setmetatable(fmem, mt)
  setmetatable(gmem, mt)
end


updatelocale()



local I = m.P(function (s,i) print(i, s:sub(1, i-1)); return i end)


local function getdef (id, Defs)
  local c = Defs and Defs[id]
  if not c then error("undefined name: " .. id) end
  return c
end


local function patt_error (s, i)
  local msg = (#s < i + 20) and s:sub(i)
                             or s:sub(i,i+20) .. "..."
  msg = ("pattern error near '%s'"):format(msg)
  error(msg, 2)
end

local function mult (p, n)
  local np = mm.P(true)
  while n >= 1 do
    if n%2 >= 1 then np = np * p end
    p = p * p
    n = n/2
  end
  return np
end

local function equalcap (s, i, c)
  if type(c) ~= "string" then return nil end
  local e = #c + i
  if s:sub(i, e - 1) == c then return e else return nil end
end


local S = (Predef.space + "--" * (any - Predef.nl)^0)^0

local name = m.R("AZ", "az") * m.R("AZ", "az", "__", "09")^0

local arrow = S * "<-"

local exp_follow = m.P"/" + ")" + "}" + ":}" + "~}" + (name * arrow) + -1

name = m.C(name)


-- identifiers only have meaning in a given environment
local Identifier = name * m.Carg(1)

local num = m.C(m.R"09"^1) * S / tonumber

local String = "'" * m.C((any - "'")^0) * "'" +
               '"' * m.C((any - '"')^0) * '"'


local defined = "%" * Identifier / function (c,Defs)
  local cat =  Defs and Defs[c] or Predef[c]
  if not cat then error ("name '" .. c .. "' undefined") end
  return cat
end

local Range = m.Cs(any * (m.P"-"/"") * (any - "]")) / mm.R

local item = defined + Range + m.C(any)

local Class =
    "["
  * (m.C(m.P"^"^-1))    -- optional complement symbol
  * m.Cf(item * (item - "]")^0, mt.__add) /
                          function (c, p) return c == "^" and any - p or p end
  * "]"

local function adddef (t, k, Defs, exp)
  if t[k] then
    error("'"..k.."' already defined as a rule")
  else
    t[k] = exp
  end
  return t
end

local function firstdef (n, Defs, r) return adddef({n}, n, Defs, r) end



local exp = m.P{ "Exp",
  Exp = S * ( m.V"Grammar"
            + m.Cf(m.V"Seq" * ("/" * S * m.V"Seq")^0, mt.__add) );
  Seq = m.Cf(m.Cc(m.P"") * m.V"Prefix"^0 , mt.__mul)
        * (#exp_follow + patt_error);
  Prefix = "&" * S * m.V"Prefix" / mt.__len
         + "!" * S * m.V"Prefix" / mt.__unm
         + m.V"Suffix";
  Suffix = m.Cf(m.V"Primary" * S *
          ( ( m.P"+" * m.Cc(1, mt.__pow)
            + m.P"*" * m.Cc(0, mt.__pow)
            + m.P"?" * m.Cc(-1, mt.__pow)
            + "^" * ( m.Cg(num * m.Cc(mult))
                    + m.Cg(m.C(m.S"+-" * m.R"09"^1) * m.Cc(mt.__pow))
                    )
            + "->" * S * ( m.Cg(String * m.Cc(mt.__div))
                         + m.P"{}" * m.Cc(nil, m.Ct)
                         + m.Cg(Identifier / getdef * m.Cc(mt.__div))
                         )
            + "=>" * S * m.Cg(Identifier / getdef * m.Cc(m.Cmt))
            ) * S
          )^0, function (a,b,f) return f(a,b) end );
  Primary = "(" * m.V"Exp" * ")"
            + String / mm.P
            + Class
            + defined
            + "{:" * (name * ":" + m.Cc(nil)) * m.V"Exp" * ":}" /
                     function (n, p) return mm.Cg(p, n) end
            + "=" * name / function (n) return mm.Cmt(mm.Cb(n), equalcap) end
            + m.P"{}" / mm.Cp
            + "{~" * m.V"Exp" * "~}" / mm.Cs
            + "{" * m.V"Exp" * "}" / mm.C
            + m.P"." * m.Cc(any)
            + name * -arrow / mm.V
            + "<" * name * ">" / mm.V;
  Definition = Identifier * arrow * m.V"Exp";
  Grammar = m.Cf(m.V"Definition" / firstdef * m.Cg(m.V"Definition")^0, adddef) /
                mm.P
}

local pattern = S * exp / mm.P * (-any + patt_error)


local function compile (p, defs)
  if mm.type(p) == "pattern" then return p end   -- already compiled
  local cp = pattern:match(p, 1, defs)
  if not cp then error("incorrect pattern", 3) end
  return cp
end

local function match (s, p, i)
  local cp = mem[p]
  if not cp then
    cp = compile(p)
    mem[p] = cp
  end
  return cp:match(s, i or 1)
end

local function find (s, p, i)
  local cp = fmem[p]
  if not cp then
    cp = compile(p)
    cp = mm.P{ mm.Cp() * cp + 1 * mm.V(1) }
    fmem[p] = cp
  end
  return cp:match(s, i or 1)
end

local function gsub (s, p, rep)
  local g = gmem[p] or {}   -- ensure gmem[p] is not collected while here
  gmem[p] = g
  local cp = g[rep]
  if not cp then
    cp = compile(p)
    cp = mm.Cs((cp / rep + 1)^0)
    g[rep] = cp
  end
  return cp:match(s)
end


-- exported names
local re = {
  compile = compile,
  match = match,
  find = find,
  gsub = gsub,
  updatelocale = updatelocale,
}

if version == "Lua 5.1" then _G.re = re end

return re
